xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+                dim_bounds[dim_bounds[:, 0] < np.min(dim_region_bounds), 0] = np.min(dim_region_bounds)  # below / left of region
+                dim_bounds[dim_bounds[:, 1] < np.min(dim_region_bounds), 1] = np.min(dim_region_bounds)  # below / left of region
+        # latitude weights scale with sine(latitude)
+        if dim in ['lat', 'latitude']:
+            dim_bounds = np.sin(np.radians(dim_bounds))
+        # weights are generally the difference of the boundaries
+        # REVIEW: Is this always true for our purposes?
+        dim_wts = np.abs(dim_bounds[:, 1] - dim_bounds[:, 0])
+        # we may want to attach the bounds to the dimension weights, too
+        dim_wts.attrs = dim_bounds.attrs
+        dim_wts.name = dim + '_wts'
+        axis_weights[dim] = dim_wts
+    # return the axis weights
+    return axis_weights
+
+
+def _combine_axis_weights(axis_weights, axis=['lat', 'lon']):
+    """Function generically re-scales axis weights for a given region.
+
+    Operations include:
+
+    - Create N-dimensional weighting array based on axis_weights and axis
+    dimensions specified
+
+    Parameters
+    ----------
+    axis_weights : dict
+        Dictionary of axis weights (using dims as keys)
+    axis : list, optional
+        List of axes that should be weighted (default ['lat', 'lon'])
+
+    Returns
+    -------
+    xr.DataArray
+        Array of weights
+
+    Examples
+    --------
+    Import and call module:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+    >>>
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+    >>> axis_weights = _get_axis_weights(tas)
+    >>> weights = _combine_axis_weights(axis_weights, axis=['lat', 'lon'])
+    """
+    # ensure averaging axis is a list
+    if type(axis) == str:
+        axis = list(axis)
+    # initialize weight matrix
+    weights = axis_weights[axis[0]]
+    if len(axis) > 1:
+        for i in np.arange(1, len(axis)):
+            weights = weights * axis_weights[axis[i]]
+    return weights
+
+
+def _averager(xvariable, weights, axis):
+    """Perform a weighted average of a dataarray.
+
+    Operations include:
+
+    - Ensure masked data receives zero weight
+    - Perform weighted average over user-specified dimensions
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``
+
+    weights : xr.DataArray
+        DataArray containing the desired weights for the variable array
+
+    axis : list
+        List of string values denoting the dims that should be averaged over
+        (e.g., axis=['lat', 'lon'])
+
+    Returns
+    -------
+    xr.DataArray
+        Dataarray that has been reduced via a weighted average.
+
+    Notes
+    -----
+    Needs to be evaluated for speed and compared to native xarray and CDAT
+    functionality.
+
+    """
+    # get masked data
+    masked = np.isnan(xvariable)
```

Comment:
This step is a huge hit on performance. It can be skipped if there are no missing values. xarray may have built-in functionality to flag this (I think for some arrays `xvariable.mask` will return an array or a boolean, but I'm not sure how bullet-proof this is). I also think it is possible to include a masked array with a dataarray, so it may be necessary to include logic to check for (and work on) masked arrays, too.

Reply:
You can extract a masked array using [xarray.DataArray.isnull](https://xarray.pydata.org/en/v0.19.0/generated/xarray.DataArray.isnull.html), which fills missing values with `np.nan`.

If we need to replace `np.nan` with another value, there is [xarray.DataArray.fillna](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.fillna.html).

> I also think it is possible to include a masked array with a dataarray, so it may be necessary to include logic to check for (and work on) masked arrays, too.

I'm not  sure if this is possible given that `DataArrays` can only store a single `np.ndarray` for its main data. The masked `DataArray` can only be stored in the coordinates, which doesn't sound right.

Reference: http://xarray.pydata.org/en/stable/generated/xarray.DataArray.to_masked_array.html